File: runtime/java/titan-ast-runtime/src/titan/ast/runtime/KeyWordDfaTokenAutomata.java
package titan.ast.runtime;

import java.io.InputStream;
import titan.ast.runtime.AstGeneratorResult.TokensResult;

/**
 * .
 *
 * @author tian wei jun
 */
class KeyWordDfaTokenAutomata extends DfaTokenAutomata {
  KeyWordAutomata keyWordAutomata;

  /**
   * 初始化oneTokenStringBuilder、eof、dfa.
   *
   * @param keyWordAutomata 关键字有限自动机
   * @param dfa 识别token的 确定有限状态自动机
   */
  KeyWordDfaTokenAutomata(KeyWordAutomata keyWordAutomata, TokenDfa dfa) {
    super(dfa);
    this.keyWordAutomata = keyWordAutomata;
  }

  @Override
  public TokensResult buildToken(InputStream byteInputStream) {
    TokensResult tokensResult = super.buildToken(byteInputStream);
    if (tokensResult.isOk()) {
      keyWordAutomata.buildToken(tokensResult.getOkData());
    }
    return tokensResult;
  }
}

File: runtime/cpp/titan-ast-runtime-lib/KeyWordDfaTokenAutomata.cpp
//
// Created by tian wei jun on 2022/11/25 0025.
//

#include "KeyWordDfaTokenAutomata.h"

KeyWordDfaTokenAutomata::KeyWordDfaTokenAutomata(
    const KeyWordAutomata *keyWordAutomata, const TokenDfa *tokenDfa)
    : DfaTokenAutomata(tokenDfa) {
  this->keyWordAutomata = keyWordAutomata;
}

// dfa delete by PersistentObject.tokenDfa
// keyWordAutomata delete by PersistentObject.keyWordAutomata
// tokens delete by caller
KeyWordDfaTokenAutomata::~KeyWordDfaTokenAutomata() = default;

TokensResult *
KeyWordDfaTokenAutomata::buildToken(const std::string *sourceFilePath) {
  auto tokensResult = DfaTokenAutomata::buildToken(sourceFilePath);
  if (tokensResult->isOk()) {
    keyWordAutomata->buildToken(tokensResult->getOkData());
  }
  return tokensResult;
}
